crates/trie-parallel/src/parallel_root.rs

```diff
@@ -0,0 +1,262 @@
+    pub fn new(view: ConsistentDbView<DB, Provider>, hashed_state: HashedPostState) -> Self {
+        Self { view, hashed_state }
+    }
+}
+
+impl<DB, Provider> ParallelStateRoot<DB, Provider>
+where
+    DB: Database,
+    Provider: DatabaseProviderFactory<DB> + Send + Sync,
+{
+    /// Calculate incremental state root in parallel.
+    pub fn incremental_root(self) -> Result<B256, ParallelStateRootError> {
+        self.calculate(false).map(|(root, _)| root)
+    }
+
+    /// Calculate incremental state root with updates in parallel.
+    pub fn incremental_root_with_updates(
+        self,
+    ) -> Result<(B256, TrieUpdates), ParallelStateRootError> {
+        self.calculate(true)
+    }
+
+    fn calculate(
+        self,
+        retain_updates: bool,
+    ) -> Result<(B256, TrieUpdates), ParallelStateRootError> {
+        let prefix_sets = self.hashed_state.construct_prefix_sets();
+        let storage_root_targets = StorageRootTargets::new(
+            self.hashed_state.accounts.keys().copied(),
+            prefix_sets.storage_prefix_sets,
+        );
+        let hashed_state_sorted = self.hashed_state.into_sorted();
+
+        // Pre-calculate storage roots in parallel for accounts which were changed.
+        debug!(target: "trie::parallel_state_root", len = storage_root_targets.len(), "pre-calculating storage roots");
+        let mut storage_roots = storage_root_targets
+            .into_par_iter()
+            .map(|(hashed_address, prefix_set)| {
+                let provider_ro = self.view.provider_ro()?;
+                let storage_root_result = StorageRoot::new_hashed(
+                    provider_ro.tx_ref(),
+                    HashedPostStateCursorFactory::new(provider_ro.tx_ref(), &hashed_state_sorted),
+                    hashed_address,
+                )
+                .with_prefix_set(prefix_set)
+                .calculate(retain_updates);
+                Ok((hashed_address, storage_root_result?))
+            })
+            .collect::<Result<HashMap<_, _>, ParallelStateRootError>>()?;
+
+        trace!(target: "trie::parallel_state_root", "calculating state root");
+        let mut trie_updates = TrieUpdates::default();
+
+        let provider_ro = self.view.provider_ro()?;
+        let hashed_cursor_factory =
+            HashedPostStateCursorFactory::new(provider_ro.tx_ref(), &hashed_state_sorted);
+        let trie_cursor_factory = provider_ro.tx_ref();
+
+        let hashed_account_cursor =
+            hashed_cursor_factory.hashed_account_cursor().map_err(ProviderError::Database)?;
+        let trie_cursor =
+            trie_cursor_factory.account_trie_cursor().map_err(ProviderError::Database)?;
+
+        let walker = TrieWalker::new(trie_cursor, prefix_sets.account_prefix_set)
+            .with_updates(retain_updates);
+        let mut account_node_iter = AccountNodeIter::new(walker, hashed_account_cursor);
+        let mut hash_builder = HashBuilder::default().with_updates(retain_updates);
```

Comment:
makes me think we may want helpers for setting these things up cuz now it's a bit verbose